### PR TITLE
refactor(orders): B2B-4623 add hidden prop to B3Table for column visibility and update tests

### PR DESCRIPTION
**Pull request by bc-veer**
Jira: [B2B-4623](https://bigcommercecloud.atlassian.net/browse/B2B-4623)

## What/Why?

This pull request refactors how table columns are conditionally displayed in the Orders page and its table component, making the logic more declarative and easier to maintain. It moves the column visibility logic into the column definitions themselves and introduces a new `hidden` property to table columns, which is then respected by the table rendering logic. Additionally, comprehensive tests have been added to verify the correct visibility of columns for different user roles and order types.

**Refactoring and simplification of column visibility logic:**

- The conditional logic for displaying columns such as "Company" and "Placed by" in the `Order` page has been moved from a filtering function to the column definitions themselves, using a new `hidden` property for each column. This makes the code more readable and maintainable. [[1]](diffhunk://#diff-a558aaee2e7eddca900f59f36caed5dd4539500e9fafea8e50ae7bab8af60788L227-R229) [[2]](diffhunk://#diff-a558aaee2e7eddca900f59f36caed5dd4539500e9fafea8e50ae7bab8af60788R245) [[3]](diffhunk://#diff-a558aaee2e7eddca900f59f36caed5dd4539500e9fafea8e50ae7bab8af60788R278-R283) [[4]](diffhunk://#diff-a558aaee2e7eddca900f59f36caed5dd4539500e9fafea8e50ae7bab8af60788L283-R292)

**Enhancements to the table component:**

- The `TableColumnItem` type in `B3Table.tsx` now includes an optional `hidden` property, allowing columns to declare whether they should be displayed.
- The `B3Table` component filters out columns with `hidden: true` before rendering both the table headers and rows, ensuring that only visible columns are shown. [[1]](diffhunk://#diff-ba8be525ef880b3e7560a6eac2b0eb89bb9cd3c7a6f5e8829999438388146fc4R158-R159) [[2]](diffhunk://#diff-ba8be525ef880b3e7560a6eac2b0eb89bb9cd3c7a6f5e8829999438388146fc4L218-R224) [[3]](diffhunk://#diff-ba8be525ef880b3e7560a6eac2b0eb89bb9cd3c7a6f5e8829999438388146fc4L250-R256)

## Rollout/Rollback

Revert the PR
## Testing

- A new test suite for the `Order` page has been added, verifying that the correct columns are visible or hidden for various user roles (B2B buyer, super admin, BC shopper) and order types (company vs non-company orders). This ensures that the new column visibility logic works as intended.

[B2B-4623]: https://bigcommercecloud.atlassian.net/browse/B2B-4623?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

---

> [!NOTE]
> **Medium Risk**
> Touches shared table rendering logic (`B3Table`) used across the storefront, so incorrect `hidden` handling could affect multiple tables; changes are mitigated by targeted Orders UI tests covering role/order-type permutations.
> 
> **Overview**
> Refactors Orders list column-visibility to be *declarative*: `Order.tsx` now defines `Company`/`Placed by` columns with a `hidden` flag (including clarified `SUPER_ADMIN` role casting and a dedicated `isSuperAdminNotAgenting` check) instead of filtering the column array.
> 
> Extends `B3Table` to support an optional `hidden` property on `TableColumnItem` and filters hidden columns out when rendering headers and row cells (non-infinite-scroll path), with a small early-return cleanup for the no-data case.
> 
> Adds a new `order.test.tsx` suite validating column visibility and key cell content across B2B buyer, super admin (agenting vs not), and B2C shopper scenarios for company vs non-company orders.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9ddd309ca3b1b044482c3b3150f095964d1256a9. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: apps/storefront/src/pages/order/Order.tsx [hunk 1]
```diff
@@ -272,6 +275,12 @@ function Order({ isCompanyOrder = false }: OrderProps) {
       render: ({ firstName, lastName }) => `${firstName} ${lastName}`,
       width: '10%',
       isSortable: true,
+      /**
+       * If the user is not a B2B user, or the user is a super admin who is not agenting,
+       * or the user is a super admin who is agenting but not on a company order, then the placed by column should be hidden.
+       */
+      hidden:
+        !isB2BUser || isSuperAdminNotAgenting || (!isSuperAdminNotAgenting && !isCompanyOrder),
```
Comment (bc-gary) [inline apps/storefront/src/pages/order/Order.tsx]: **[Required]** — Redundant boolean clause

`A || (!A && B)` simplifies to `A || B`, so this entire expression reduces to:

```ts
hidden: !isB2BUser || isSuperAdminNotAgenting || !isCompanyOrder,
```

The `!isSuperAdminNotAgenting &&` guard adds no logical value because the `isSuperAdminNotAgenting` case is already handled by the preceding disjunct.

The comment is also misleading — it says the third clause applies to "a super admin who is agenting" but it actually covers *any* user not on a company order.

Suggest extracting as a named boolean for clarity:
```ts
const shouldHidePlacedBy = !isB2BUser || isSuperAdminNotAgenting || !isCompanyOrder;
```

File: apps/storefront/src/pages/order/Order.tsx [hunk 2]
```diff
@@ -224,7 +224,9 @@ function Order({ isCompanyOrder = false }: OrderProps) {
     });
   };
 
-  const columnAllItems = [
+  const isSuperAdminNotAgenting = role === CustomerRole.SUPER_ADMIN && !isAgenting;
+
+  const columnItems: TableColumnItem<ListItem>[] = [
```
Comment (bc-gary) [inline apps/storefront/src/pages/order/Order.tsx]: **[Blocker]** — Merge conflict with PR #850 (B2B-4616)

PR #850 (`b2b-4616`) heavily modifies the same area of `Order.tsx` — it replaces `columnAllItems` + `getColumnItems()` with three-path branching for unified company orders, adds `useCompanyOrdersState`, `fetchUnifiedCompanyOrders`, etc.

Both PRs remove `getColumnItems()` and restructure the column definitions. When both merge to `main`, there will be non-trivial merge conflicts.

This PR should either be rebased onto `b2b-4616` or merged after #850 with a follow-up rebase to resolve conflicts.

File: apps/storefront/src/pages/order/order.test.tsx [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+describe('Orders list (UI)', () => {
+  beforeEach(() => {
+    vi.spyOn(ordersApi, 'getB2BAllOrders').mockResolvedValue(listResponse);
+    vi.spyOn(ordersApi, 'getBCAllOrders').mockResolvedValue(listResponse);
+    vi.spyOn(ordersApi, 'getOrderStatusType').mockResolvedValue(orderStatusRows);
+    vi.spyOn(ordersApi, 'getBcOrderStatusType').mockResolvedValue(orderStatusRows);
+    vi.spyOn(ordersApi, 'getCreatedByUserForOrders').mockResolvedValue({
+      createdByUser: { results: [] },
+    });
+  });
```
Comment (bc-gary) [inline apps/storefront/src/pages/order/order.test.tsx]: **[Required]** — Uses `vi.spyOn` instead of MSW (project convention)

All existing order tests (`unified-orders.test.tsx`, `CompanyOrderList/index.test.tsx`, `CompanyOrderList/index.mobile.test.tsx`) use MSW's `graphql.query()` handlers to intercept requests at the network level.

This test uses `vi.spyOn(ordersApi, 'getB2BAllOrders').mockResolvedValue(...)` which bypasses the actual GraphQL query construction and is inconsistent with the rest of the order test suite.

Spike decision (B2B-4712): "Functional tests use MSW at component level (not service-level unit tests)."

File: apps/storefront/src/pages/order/order.test.tsx [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+        },
+      },
+    });
+
+    await waitFor(() => {
+      expect(screen.getByTestId('tableHead-orderId')).toBeVisible();
+    });
+    expect(screen.getByTestId('tableHead-placedBy')).toBeVisible();
+  });
+});
```
Comment (bc-gary) [inline apps/storefront/src/pages/order/order.test.tsx:359]: **[Suggestion]** — Missing edge case: super admin agenting on My Orders

The tests cover super admin agenting on Company Orders (placedBy visible) and super admin NOT agenting on Company Orders (placedBy hidden). But there's no test for **super admin who IS agenting viewing My Orders** — `placedBy` should be hidden because `!isCompanyOrder` is true.

This edge case validates that the `!isCompanyOrder` clause works correctly even when the user has agenting privileges.

File: apps/storefront/src/pages/order/Order.tsx (hunk 2 above)
Comment (bc-veer) [reply apps/storefront/src/pages/order/Order.tsx]: Happy to wait for #850 to be merged

File: apps/storefront/src/pages/order/Order.tsx (hunk 1 above)
Comment (bc-veer) [reply apps/storefront/src/pages/order/Order.tsx]: Great. Thank you.

[the remark this answers is not in the document]
Comment (bc-veer) [reply apps/storefront/src/pages/order/Order.tsx]: Great. Thank you.

[the remark this answers is not in the document]
Comment (bc-veer) [reply apps/storefront/src/pages/order/Order.tsx]: Great. Thank you.

File: apps/storefront/src/pages/order/order.test.tsx (hunk 3 above)
Comment (bc-veer) [reply apps/storefront/src/pages/order/order.test.tsx]: Updated

File: apps/storefront/src/pages/order/order.test.tsx (hunk 4 above)
Comment (bc-veer) [reply apps/storefront/src/pages/order/order.test.tsx:359]: Added new test for edge case

File: apps/storefront/src/pages/order/table/B3Table.tsx [hunk 5]
```diff
@@ -151,144 +155,145 @@ export function B3Table<Row extends OrderIdRow>({
     });
   };
 
-  return listItems.length > 0 ? (
```
Comment (bc-veer) [inline apps/storefront/src/pages/order/table/B3Table.tsx:154]: Replaced with early return to avoid nested conditions.

**Review by bc-gary — APPROVED** [review]
Nice work @bc-veer 👏 

All four previous review comments have been addressed:

1. **Redundant boolean clause** — simplified to `!isB2BUser || isSuperAdminNotAgenting || !isCompanyOrder`
2. **Merge conflict with PR #850** — rebased onto latest main
3. **vi.spyOn → MSW** — tests now use `startMockServer()` + `graphql.query()` handlers
4. **Missing edge case** — added test for super admin agenting on My Orders

The `hidden` property approach is clean, the B3Table early-return restructure improves readability, and test coverage is comprehensive across all role/order-type combinations. LGTM.

File: apps/storefront/src/pages/order/order.test.tsx [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+            },
+          },
+        }),
+      ),
+    );
+
+    renderWithProviders(<Order isCompanyOrder />, { preloadedState: basePreloaded });
+
+    await waitFor(() => {
+      expect(screen.getByTestId('tableHead-orderId')).toBeVisible();
```
Comment (bc-lulu-yang) [inline apps/storefront/src/pages/order/order.test.tsx]: [suggestion]

Generally based on React Testing Library, we should prefer queries that match how users find things: roles, labels, and visible text, which is I personally love to follow.

For example, we can use below instead: 
`expect(screen.getByRole('columnheader', { name: 'Order' })).toBeVisible();`

**Review by bc-lulu-yang — APPROVED** [review]
LGTM, with only one suggestion: generally in React Testing Library we should prefer queries that match how users find things: roles, labels, and visible text

File: apps/storefront/src/pages/order/Order.tsx [hunk 7]
```diff
@@ -287,80 +287,70 @@ function Order({ isCompanyOrder = false }: OrderProps) {
...
-      render: ({ orderId }) => orderId,
-    },
-    {
-      key: 'companyName',
-      title: b3Lang('orders.company'),
-      width: '10%',
-      isSortable: false,
-      render: ({ companyInfo }) => {
-        return <Box>{companyInfo?.companyName || '–'}</Box>;
+  const isSuperAdminNotAgenting = role === CustomerRole.SUPER_ADMIN && !isAgenting;
```
Comment (bc-jeffteng) [inline apps/storefront/src/pages/order/Order.tsx]: [Question]
I noticed that in the deleted code, we use `Number(role)` (in delete code ln:352), why we change back to only `role`. B2B user role can be stringly typed, so `'3' === CustomerRole.SUPER_ADMIN` would miss super admins.

Comment (bc-veer) [reply apps/storefront/src/pages/order/Order.tsx]: Thank you @bc-jeffteng  for notifying us about the change.

I’ve safeguarded the role comparison using Number() to ensure the logic works correctly.

File: apps/storefront/src/pages/order/order.test.tsx (hunk 6 above)
Comment (bc-veer) [reply apps/storefront/src/pages/order/order.test.tsx]: Updated. Thank you @bc-lulu-yang

**Comment by bc-launchbay** [comment]
*b2b-buyer-portal* deployed to Staging GCS Upload
success [20260512064736](https://launchbay.bigcommerce.net/projects/515/releases/284438) by Launchbay

**Comment by bc-launchbay** [comment]
*b2b-buyer-portal* deployed to Production GCS Upload
success [20260512064736](https://launchbay.bigcommerce.net/projects/515/releases/284438) by Launchbay
